app/upload.go

```diff
@@ -19,6 +19,67 @@ import (
 
 const minFirstPartSize = 5 * 1024 * 1024 // 5MB
 
+func (a *App) runPluginsHook(info *model.FileInfo, file io.Reader) *model.AppError {
+	pluginsEnvironment := a.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return nil
+	}
+
+	// using a pipe to avoid loading the whole file content in memory.
+	r, w := io.Pipe()
+	errChan := make(chan *model.AppError, 1)
```

Comment:
It's not immediately apparent to me why we need a buffered channel. The write happens in a separate goroutine, so even if we haven't started to read yet, we will at some point anyways.

Sorry, I do realize that this is not new code and you just moved it to a new file. Nevertheless, since you are making a great overhaul here, maybe it's good to look into these as well.

Reply:
> It's not immediately apparent to me why we need a buffered channel. The write happens in a separate goroutine, so even if we haven't started to read yet, we will at some point anyways.

I believe the main reason was to make the send non-blocking.

> Sorry, I do realize that this is not new code and you just moved it to a new file. Nevertheless, since you are making a great overhaul here, maybe it's good to look into these as well.

This code is fairly new actually. And it's all my doing so better to keep it sane :stuck_out_tongue:

Reply:
> I believe the main reason was to make the send non-blocking.

Gotcha, see my comment below.